File: src/main/java/net/theevilreaper/vulpes/generator/generation/Generator.java
package net.theevilreaper.vulpes.generator.generation;

import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;

/**
 * The {@link Generator} interface defines the base method structure which each implementations needs to work with the current system approach.
 *
 * @author theEvilReaper
 * @version 1.0.0
 * @since 0.1.0
 */
public interface Generator {

    /**
     * Each generator must implement his own logic for this method
     */
    void generate(@NotNull Path javaPath);

    /**
     * Returns the name of the generator
     */
    @NotNull String getName();

    /**
     * Cleanup the generator
     */
    void cleanup();
}

File: src/main/kotlin/net.theevilreaper.vulpes.generator/generation/Generator.kt
package net.theevilreaper.vulpes.generator.generation

import net.theevilreaper.vulpes.generator.generation.type.GeneratorType
import java.nio.file.Path

/**
 * The interface defines some basic structure for generators
 * @author Joltras
 * @version 1.0.0
 * @author 1.0.0
 */
interface Generator {

    /**
     * Each generator must implement his own logic for this method
     */
    fun generate(javaPath: Path)

    /**
     * Returns the name from the generator.
     * @return the given generator name
     */
    fun getName(): String

    /**
     * Contains logic to clean up the generator data structure.
     */
    fun cleanUp()

    /**
     * Returns the given [GeneratorType] from the generator implementation.
     * @return the given type
     */
    fun getType(): GeneratorType
}
